File: RandomCellCoordinateGenerator.lua
RandomCellCoordinateGenerator = {}

function RandomCellCoordinateGenerator:new(grid)
    local o = {}
    setmetatable(o, self)
    self._grid = grid
    self.__index = self
    math.randomseed(os.time())
    return o
end

function RandomCellCoordinateGenerator:generate()
    local row, column
    repeat
        local randValue = math.random(0, GRID_SIZE * GRID_SIZE - 1)
        row = math.floor(randValue / GRID_SIZE)
        column = randValue - row * GRID_SIZE
    until (self._grid:isValidCell(row, column))
    return row, column
end

File: RandomCellCoordinateGenerator.cpp
#include <random>
#include "RandomCellCoordinateGenerator.h"

std::pair<short, short> RandomCellCoordinateGenerator::generate() {
    short row, column;
    do {
        int randValue = m_distribution(m_generator);
        row = (short) (randValue / Grid::SIZE);
        column = (short) (randValue - row * Grid::SIZE);
    } while (!m_grid.isValidCell(row, column));
    return {row, column};
}

void RandomCellCoordinateGenerator::init() {
    std::random_device rd;
    m_generator = std::mt19937(rd());
    m_distribution = std::uniform_int_distribution<>(0, Grid::SIZE * Grid::SIZE - 1);
}
